### ISSUE TITLE
update Context.ClientIP() to use standard github.com/sebest/xff/ package

### ISSUE THREAD

**Issue by oryband** [gin-gonic/gin#337]
following discussion in gin-gonic/contrib#49 , i think maybe we should use https://github.com/sebest/xff instead of implementing header parsing? This package looks like it does the job better, filtering private ip addresses and such: https://github.com/sebest/xff/blob/master/xff.go#L11

**Comment by slimmy** [comment]
Interesting idea but I don't think panics in non-initialization code is a good pattern (https://github.com/sebest/xff/blob/master/xff.go#L14)

**Comment by manucorporat** [comment]
@slimmy  I think that code is part of initialization data and in fact they use a panic() in the same way an [assert](http://www.cplusplus.com/reference/cassert/assert/) is used in C. ie. a defensive programming practice, there is no way panic() will be called.

**Comment by slimmy** [comment]
@manucorporat you're right, I read the code too quickly

**Comment by oryband** [comment]
so, @manucorporat, lets merge this

**Comment by manucorporat** [comment]
There are several parts of the xff implementation that doesn't look correct.

**Comment by oryband** [comment]
Which ones?

**Comment by manucorporat** [comment]
- Iterating over all the ForwardIP values, I think the first one is the only important one.

    ```go
	for _, ip := range strings.Split(ipList, ",") {
		ip = strings.TrimSpace(ip)
```
- It returns the port (a fake port) since the original client probably does not connect using that port. net.JoinHostPort(ip, oport)
- If IP is not public, we return "" ?

**Comment by oryband** [comment]
@manucorporat you are absoloutely right, closing this issue

**Comment by sebest** [comment]
@manucorporat a few comments on yours:
 * the first IP is not always the most important especially if you have more than one, it is often the one that you can trust the less as it can be forged, this can be an issue with Amazon Elastic Load Balancer as it does not do any validation, see  http://docs.aws.amazon.com/ElasticLoadBalancing/latest/DeveloperGuide/x-forwarded-headers.html#x-forwarded-for
 * The first IP can be a private IP if the first proxy is a private proxy.
 * Nginx real_ip module can configure the behaviour: http://nginx.org/en/docs/http/ngx_http_realip_module.html#real_ip_recursive (xff could add this feature)
 * About the port number, you can not reliably get the port number with X-Forwarded-For so xff does not even try, there is a specific Header for that purpose: X-Forwarded-Port
 * If IP in the X-Forwarded-For header is not public , xff just does not change the remote IP it does not return "" https://github.com/sebest/xff/blob/master/xff.go#L54 , again it is a choice of implementation as in most use case you need X-Forwarded-For to get a public IP, this behavior could also be configurable, but i think that if you expect Private IPs it is either a network that you control and you should implement your own logic for parsing X-Forwarded-For or if it is a network that you don't control, you would not be able to make any sense of the private IPs, so having the public IPs of the proxy in front of these private IPs makes more sense.
To sum up, the way you parse the X-Forwarded-For header really depends on your use case. This one is geared toward GeoIP with no security concern.

**Comment by oryband** [comment]
@bolshoy

**Comment by oryband** [comment]
@manucorporat many implement @sebest 's ideas as well, and provide an option to choose between the two?
